### PR TITLE
HTTPDownloader

### PR DESCRIPTION
**Pull request by Saurav-D**
# Peer Review Information

I've added:
- `HTTPDownloader` class which extends `DatasetDownloader` class
- `checksum_file` option in the download cli.

# Pull Request Check List

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: datasetinsights/io/downloader/http_downloader.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    @staticmethod
+    def get_checksum_from_file(filepath):
+
+        if filepath.startswith(("http://", "https://")):
+            with tempfile.TemporaryDirectory() as tmp:
+                checksum_file_path = os.path.join(tmp, "checksum.txt")
+                download_file(source_uri=filepath, dest_path=checksum_file_path)
+                return HTTPDownloader.read_checksum_from_txt(checksum_file_path)
+
+        elif filepath.startswith("/"):
```
Comment (AdamPalmarUnity) [inline datasetinsights/io/downloader/http_downloader.py]: You might want to do `os.path.isfile(fname) ` instead. It's possible that the user specifies a relative path to the file based on where the user is executing the command.

Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Thanks, will do.

File: datasetinsights/io/downloader/http_downloader.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import tempfile
+import zipfile
+
+from datasetinsights.io.download import download_file, validate_checksum
+from datasetinsights.io.downloader.base import DatasetDownloader
+from datasetinsights.io.exceptions import ChecksumError
+
+logger = logging.getLogger(__name__)
+
+DATASETS = ["synthetic", "groceries_real"]
```
Comment (AdamPalmarUnity) [inline datasetinsights/io/downloader/http_downloader.py]: What will these variables be used for?

Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Forgot to remove them. Will remove.

File: datasetinsights/io/downloader/http_downloader.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    def get_checksum_from_file(filepath):
+
+        if filepath.startswith(("http://", "https://")):
+            with tempfile.TemporaryDirectory() as tmp:
+                checksum_file_path = os.path.join(tmp, "checksum.txt")
+                download_file(source_uri=filepath, dest_path=checksum_file_path)
+                return HTTPDownloader.read_checksum_from_txt(checksum_file_path)
+
+        elif os.path.isfile(filepath):
+            checksum_file_path = filepath
```
Comment (AdamPalmarUnity) [inline datasetinsights/io/downloader/http_downloader.py]: [minor] pass the filepath directly to the staticmethod

File: datasetinsights/datasets/groceries_real.py [hunk 4]
```diff
@@ -126,19 +110,13 @@ def __init__(
...
-                f"Available versions are: {valid_versions}"
-            )
         logger.info(
             f"Using split {split} and version {version} for groceries real "
             "dataset."
         )
 
         self.version = version
-        self.root = os.path.join(data_root, self.LOCAL_PATH)
+        self.root = data_root
```
Comment (AdamPalmarUnity) [inline datasetinsights/datasets/groceries_real.py]: I think it makes sense to rename this to data_path or data_directory or something along those lines. As well as the data_root constructor argument

File: tests/test_http_downloader.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import pytest
+
+from datasetinsights.io.downloader.http_downloader import HTTPDownloader
+from datasetinsights.io.exceptions import ChecksumError
+
+
+def test_read_checksum_from_txt():
+    with tempfile.NamedTemporaryFile(mode="w+") as tmp:
+        tmp.write("123456")
+        tmp.read()
```
Comment (AdamPalmarUnity) [inline tests/test_http_downloader.py]: I think it makes more sense to do a tmp.flush() to force the write to the file instead of tmp.read()

File: tests/test_http_downloader.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        assert HTTPDownloader.get_checksum_from_file(tmp.name) == 123456
+
+
+@patch("datasetinsights.io.downloader.http_downloader.download_file")
+@patch.object(HTTPDownloader, "read_checksum_from_txt")
+def test_get_checksum_from_file(
```
Comment (AdamPalmarUnity) [inline tests/test_http_downloader.py]: Unit test should follow the pattern. 
arrange 
act 
assert 

This will make the tests easier to understand. This unit test does multiple things and should instead be refactored into multiple tests.

File: tests/test_http_downloader.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    mock_read_checksum_from_txt, mock_download_file
+):
+    filepath_1 = "http://some/path"
+    filepath_2 = "https://some/path"
+    filepath_3 = "some/wrong/path"
+
+    mock_read_checksum_from_txt.return_value = 123456
```
Comment (AdamPalmarUnity) [inline tests/test_http_downloader.py:26]: This defeats the purpose of the test. Mocking the function you are testing does not validate the behavior it just tests if mocking a function works.

File: tests/test_http_downloader.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    HTTPDownloader.get_checksum_from_file(filepath_1)
+    mock_download_file.assert_called_once()
+    mock_read_checksum_from_txt.assert_called_once()
+
+    assert HTTPDownloader.get_checksum_from_file(filepath_2) == 123456
+
+    with pytest.raises(ValueError):
+        HTTPDownloader.get_checksum_from_file(filepath_3)
+
+    with tempfile.NamedTemporaryFile() as tmp:
```
Comment (AdamPalmarUnity) [inline tests/test_http_downloader.py]: This case is already being tested in test_read_checksum_file_from_txt

File: datasetinsights/io/downloader/base.py [hunk 9]
```diff
@@ -19,15 +19,16 @@ def _find_downloader(source_uri):
     """
     protocols = "|".join(_registry.keys())
     pattern = re.compile(f"({protocols})")
+
     protocol = pattern.findall(source_uri)
-    if protocol:
+
+    if source_uri.startswith(("https://", "http://")):
```
Comment (adason) [inline datasetinsights/io/downloader/base.py:24]: If protocol can't be matched, this will fail. `None string have no method called start_with...`. This should be fixed and add unittest tests for such cases.

File: datasetinsights/io/downloader/base.py [hunk 10]
```diff
@@ -48,6 +49,9 @@ def create_downloader(source_uri, **kwargs):
 
 
 class DatasetDownloader(ABC):
+    def __init__(self, **kwargs):
```
Comment (adason) [inline datasetinsights/io/downloader/base.py:51]: Should be removed. This is not required.

File: datasetinsights/io/downloader/http_downloader.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+logger = logging.getLogger(__name__)
+
+
+class HTTPDownloader(DatasetDownloader, protocol="http://"):
+    """ This class is used to download data from any HTTP or HTTPS public url
+        and perform function such as downloading the dataset and checksum
+        validation.
+    """
+
+    def __init__(self, **kwargs):
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: Not required to have __init__ method.

File: datasetinsights/commands/download.py [hunk 12]
```diff
@@ -79,8 +79,15 @@ def convert(self, value, param, ctx):
     help="Unity Simulation access token. "
     "This will override synthetic datasets source-uri for Unity Simulation",
 )
+@click.option(
+    "--checksum-file",
+    type=str,
+    default=None,
+    help="Dataset checksum text file. "
```
Comment (adason) [inline datasetinsights/commands/download.py]: Since `checksum-file` can either be a http url or file in the local filesystem, you should make this clear in the help message.

File: datasetinsights/io/downloader/http_downloader.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+                validate_checksum(dataset_path, checksum)
+            except ChecksumError as e:
+                logger.info("Checksum mismatch. Deleting the downloaded file.")
+                os.remove(dataset_path)
+                raise e
+
+        HTTPDownloader.unzip_file(dataset_path, output)
+
+    @staticmethod
+    def get_checksum_from_file(filepath):
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: This method seems to be generic enough that can be used for other `DatasetDownloader`. Maybe it should be moved to the same module where we compute checksums?

File: datasetinsights/io/downloader/http_downloader.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+                return HTTPDownloader.read_checksum_from_txt(checksum_file_path)
+
+        elif os.path.isfile(filepath):
+            return HTTPDownloader.read_checksum_from_txt(filepath)
+
+        else:
+            raise ValueError(f"Can not get checksum from path: {filepath}")
+
+    @staticmethod
+    def read_checksum_from_txt(filepath):
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: This method seems to be generic enough that can be used for other `DatasetDownloader`. Maybe it should be moved to the same module where we compute checksums?

File: datasetinsights/io/downloader/http_downloader.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+            checksum = HTTPDownloader.get_checksum_from_file(checksum_file)
+            try:
+                logger.debug("Validating checksum!!")
+                validate_checksum(dataset_path, checksum)
+            except ChecksumError as e:
+                logger.info("Checksum mismatch. Deleting the downloaded file.")
+                os.remove(dataset_path)
+                raise e
+
+        HTTPDownloader.unzip_file(dataset_path, output)
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: [nit] use `self.unzip_file(...)` to save some typing. This will also make it easier if we have to change the class name `HTTPDownloader` in the future.

File: datasetinsights/io/downloader/http_downloader.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+                                 can be HTTP or HTTPS url or local path.
+
+        Raises:
+            ChecksumError: This will raise this error if checksum doesn't
+                           matches
+
+        """
+
+        dataset_path = os.path.join(output, "dataset.zip")
+        download_file(source_uri, dataset_path)
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py:42]: What happened if the file download was not successful? Should we catch error and re-raise exceptions?

File: tests/test_create_downloader.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+from datasetinsights.io.downloader.unity_simulation import (
+    UnitySimulationDownloader,
+)
+
+
+@pytest.mark.parametrize(
+    "source_uri", ["http://", "https://"],
+)
+def test_create_downloader_http_downloader(source_uri):
+    # arrange
```
Comment (adason) [inline tests/test_create_downloader.py]: ??? This comment seems out of place.

File: tests/test_create_downloader.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+    # act
+    downloader = create_downloader(source_uri=source_uri)
+
+    # assert
+    assert isinstance(downloader, HTTPDownloader)
+
+
+def test_create_downloader_unity_simulation_downloader():
```
Comment (adason) [inline tests/test_create_downloader.py:22]: Why test create_downloader when `_find_downloader` is tested against different cases? What about cases when _find_downloader can find a downloader for a given source_uri? What about cases when source_uri is none?

File: tests/test_http_downloader.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    )
+
+    # assert
+    mock_download_file.assert_called_once()
+    mock_unzip.assert_called_once_with(dataset_path, output)
+    mock_get_checksum_from_file.assert_called_once()
+    mock_validate_check_sum.assert_called_once()
+
+
+@pytest.mark.parametrize(
```
Comment (adason) [inline tests/test_http_downloader.py]: Why do we need different parameter when we have `mock_validate_checksum.side_effect = ChecksumError`. In this does it matter to vlaidate multiple combination of `source_uri` and `checksum_file` here, if `validate_checksum` method is always going to raise?

File: tests/test_http_downloader.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        HTTPDownloader.get_checksum_from_file(filepath)
+
+    # assert
+    mock_download_file.assert_not_called()
+    mock_read_checksum_from_txt.assert_not_called()
+
+
+@pytest.mark.parametrize(
+    "source_uri",
+    ["http://some/path", "https://some/path", "some/checksum_file.txt"],
```
Comment (adason) [inline tests/test_http_downloader.py]: Why do you need a `some/checksum_file.txt` for `source_uri`? Also, the name of this method seems to suggest it tries to run logic without checksum. Why do we need to have both `source_uri=http, https` when the logic of handling either http/https source uri is already covered by different use cases? Ideally, user should be able to provide any random string for `source_uri` and be able to test the behavior when `checksum_file` is not provided.

File: tests/test_http_downloader.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+    with tempfile.NamedTemporaryFile(mode="w+") as tmp:
+        tmp.write("123456")
+        tmp.flush()
+        # act
+        checksum = HTTPDownloader.get_checksum_from_file(tmp.name)
+        # assert
+        assert checksum == 123456
+
+
+@pytest.mark.parametrize("filepath", ["http://some/path", "https://some/path"])
```
Comment (adason) [inline tests/test_http_downloader.py]: It will be nice to add  a few test cases when a given `filepath` will raise `ValueError`.

File: datasetinsights/io/downloader/http_downloader.py (hunk 16 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py:42]: That's been handled in `download_file` function. It'll raise `DownloadError` if unsuccessful.

File: datasetinsights/io/downloader/http_downloader.py (hunk 11 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/http_downloader.py]: This will fail if the constructor is called with an argument like access_token.

File: tests/test_create_downloader.py (hunk 17 above)
Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py]: The arrange part could be removed

File: tests/test_create_downloader.py (hunk 18 above)
Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py:22]: Good point, there is no test for create_downloader with input None. I don't understand what you mean by the two first questions though.

File: tests/test_http_downloader.py (hunk 21 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: The test below does this.

File: tests/test_http_downloader.py (hunk 20 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: The some/checksum_file.txt can be removed from the test case. The test is about not providing a checksum file and validating that the download and unzip is called properly

File: tests/test_http_downloader.py (hunk 19 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: The importance of the test is the following

```
    mock_get_checksum_from_file.assert_called_once()
    mock_download_file.asert_called_once()
    mock_remove.assert_called_once()
    mock_unzip_file.assert_not_called()
```

File: tests/test_http_downloader.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+    # act
+    with pytest.raises(ChecksumError):
+        downloader.download(
+            source_uri=source_uri, output=output, checksum_file=checksum_file
+        )
+
+    # assert
+    mock_get_checksum_from_file.assert_called_once()
+    mock_download_file.asert_called_once()
```
Comment (AdamPalmarUnity) [inline tests/test_http_downloader.py]: Typo here.

Comment (Saurav-D) [reply tests/test_http_downloader.py]: What's the typo?

File: datasetinsights/io/downloader/base.py (hunk 9 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/base.py:24]: Are you sure. its the source_uri that is being called not protocol

File: tests/test_http_downloader.py (hunk 20 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: I don't understand your comment on what is wrong with the test besides some/checksum_file.txt

File: datasetinsights/io/downloader/http_downloader.py (hunk 13 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Moved to `download.py` where checksum is computed.

File: datasetinsights/io/downloader/http_downloader.py (hunk 14 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Moved to `download.py` where checksum is computed.

File: tests/test_http_downloader.py (hunk 20 above)
Comment (Saurav-D) [reply tests/test_http_downloader.py]: I think @adason mean that we don't need to test for both `http` and `https` since this test doesn't check whether it handles `http` and `https` but it's about whether download and unzip is called properly.

File: tests/test_http_downloader.py [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+    # act
+    downloader.download(source_uri=source_uri, output=output)
+
+    # assert
+    mock_download_file.assert_called_once()
+    mock_unzip.assert_called_once_with(dataset_path, output)
+
+
+@pytest.mark.parametrize(
```
Comment (Saurav-D) [inline tests/test_http_downloader.py]: Here also we don't need to parameterize `source_uri`

File: tests/test_http_downloader.py (hunk 22 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: asert_called_once

File: tests/test_http_downloader.py (hunk 20 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: Ok in that case we can remove the parametrize method and just have one source_uri

File: datasetinsights/io/downloader/base.py (hunk 10 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/base.py:51]: I'll remove it from here

File: tests/test_create_downloader.py (hunk 18 above)
Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py:22]: I don't think it makes sense to unit test _find_downloader as it is a private function

File: tests/test_http_downloader.py (hunk 19 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: I'll remove the parametrize for this test

File: tests/test_http_downloader.py (hunk 21 above)
Comment (AdamPalmarUnity) [reply tests/test_http_downloader.py]: Are there any cases you have in mind?

File: datasetinsights/io/downloader/base.py (hunk 9 above)
Comment (adason) [reply datasetinsights/io/downloader/base.py:24]: You are correct. 👍

File: datasetinsights/io/downloader/http_downloader.py (hunk 11 above)
Comment (adason) [reply datasetinsights/io/downloader/http_downloader.py]: `__init__` method should lookup it's parent class `DatasetDownloader` which should be able to catch access_token in  **kwargs?

File: datasetinsights/io/downloader/http_downloader.py [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+                validate_checksum(dataset_path, checksum)
+            except ChecksumError as e:
+                logger.info("Checksum mismatch. Deleting the downloaded file.")
+                os.remove(dataset_path)
+                raise e
+
+        self.unzip_file(dataset_path, output)
+
+    @staticmethod
+    def unzip_file(filepath, destination):
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: @AdamPalmarUnity @Saurav-D 

I just realized that if we push the responsibility of unzipping files to `DatasetDownloader`, we actually make a very strong assumption about the file that should be downloaded. In fact, don't know what type of compressed file we have for a given `source_uri`. It could download an tfdata `protobuf` file, `hdf5` file, `tar.gz` file, ... 

**We should not make unzip part of the responsibility of HTTPDownloader**

File: tests/test_create_downloader.py (hunk 18 above)
Comment (adason) [reply tests/test_create_downloader.py:22]: We should have unittest for ANY function that preform non-trivial operations, whether it's public or private function. The logic inside `create_downloader` seems trivial however `_find_downloader` is non trivial.  

Typo in my previous comment. Correction: 

> What about cases when `_find_downloader` can't find a `DatasetDownloader` from a given source_uri? What about cases when source_uri is none?

File: datasetinsights/io/downloader/base.py (hunk 10 above)
Comment (adason) [reply datasetinsights/io/downloader/base.py:51]: Actually, from the comments in `HTTPDownloader`, we should keep this to catch all arguments that was passed to constructor.  By default, when users creates a new Downloader without `__init__` method, it's parent class `DatasetDownloader` should handle such cases.

File: tests/test_http_downloader.py (hunk 21 above)
Comment (adason) [reply tests/test_http_downloader.py]: A few invalid cases: ["zvsssdfsdg", ":path with space", ""]

Comment (Saurav-D) [reply tests/test_http_downloader.py]: Added [here](https://github.com/Unity-Technologies/datasetinsights/pull/47/commits/93e30c6571d0fb0db175dacc03e7cfb072fc5a0e)

File: datasetinsights/io/downloader/http_downloader.py (hunk 24 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Then whose responsibility should be to unzip the dataset file. Since Dataset class expects just a data_directory which contains all the files (annotations, images etc.) when should unzip happen and where? Also, do we want to support different dataset file types?

File: datasetinsights/io/downloader/http_downloader.py [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+            checksum = get_checksum_from_file(checksum_file)
+            try:
+                logger.debug("Validating checksum!!")
+                validate_checksum(dataset_path, checksum)
+            except ChecksumError as e:
+                logger.info("Checksum mismatch. Deleting the downloaded file.")
+                os.remove(dataset_path)
+                raise e
+
+        self.unzip_file(dataset_path, output)
```
Comment (BlairLee) [inline datasetinsights/io/downloader/http_downloader.py]: [must fix] I've merged this branch to my local branch. I figured out that this would unzip it to a subfolder. 
For example, if we set output = "/data/groceries/v3". It would download a "dataset.zip" in the "/data/groceries/v3" and unzip it to "/data/groceries/v3/v3".

Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: Yes, that's because groceries dataset zip file contains a directory `v3` which in turn contains the dataset files. I don't think it should be the responsibility of downloader to find the files or get rid of directories, instead all the dataset files should be zipped directly instead of a zipping a directory that contains those files.

Comment (BlairLee) [reply datasetinsights/io/downloader/http_downloader.py]: I will zip the GroceriesReal directly and update on GCS.

File: datasetinsights/io/downloader/http_downloader.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        """ This method is used to download the dataset from HTTP or HTTPS url.
+
+        Args:
+            source_uri (str): This is the downloader-uri that indicates where
+                              the dataset should be downloaded from.
+
+            output (str): This is the path to the directory where the download
+                          will store the dataset.
+
+            checksum_file (str): This is path of the txt file that contains
```
Comment (BlairLee) [inline datasetinsights/io/downloader/http_downloader.py:31]: Are we going to provide checksum files of the datasets in the public GCS bucket?

File: datasetinsights/io/downloader/base.py (hunk 10 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/base.py:51]: I'll keep it here then and remove it from the other http_downloader

File: tests/test_create_downloader.py (hunk 18 above)
Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py:22]: Added these two tests that test the interface.


```
def test_create_downloader_invalid_input():
    # arrange
    source_uri = "invalid_protocol://"
    # assert
    with pytest.raises(ValueError):
        # act
        create_downloader(source_uri=source_uri)


def test_create_downloader_none_input():
    # arrange
    source_uri = None
    # assert
    with pytest.raises(TypeError):
        # act
        create_downloader(source_uri=source_uri)
```

Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py:22]: I think we should focus on testing the interface instead of the implementation

File: datasetinsights/io/downloader/http_downloader.py (hunk 24 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/http_downloader.py]: In that case i think it makes sense to refactor the unzipping/ decompressing into another class that is injected into this class. Then we can also reuse that in the GCSDownloader. We could call it decompressor. It could also be a function.

File: tests/test_create_downloader.py (hunk 18 above)
Comment (AdamPalmarUnity) [reply tests/test_create_downloader.py:22]: What kind of cases is missing?

File: datasetinsights/io/downloader/http_downloader.py (hunk 24 above)
Comment (AdamPalmarUnity) [reply datasetinsights/io/downloader/http_downloader.py]: To answer Saurav's question i still think the unzipping/decompressing can happen in the HTTPDownloader but the code for how to do it should not live here.

File: datasetinsights/io/downloader/http_downloader.py (hunk 26 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py:31]: Yes.

File: datasetinsights/io/compression.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+
+        """
+        destination = os.path.join(
+            destination, os.path.splitext(os.path.basename(filepath))[0]
+        )
+        with open(destination, "wb") as out_f, gzip.GzipFile(filepath) as zip_f:
+            out_f.write(zip_f.read())
+
+
+def get_file_type_from_filepath(filepath):
```
Comment (AdamPalmarUnity) [inline datasetinsights/io/compression.py]: Would probably make this private.

**Review by AdamPalmarUnity — DISMISSED** [review]

File: tests/test_compression.py [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+
+from datasetinsights.io.compression import (
+    GZipCompression,
+    TarFileCompression,
+    ZipFileCompression,
+    compression_factory,
+)
+
+
+@patch("datasetinsights.io.compression._get_file_extension_from_filepath")
```
Comment (AdamPalmarUnity) [inline tests/test_compression.py]: Why are you mocking out the private function?

Comment (Saurav-D) [reply tests/test_compression.py]: Because there is no actual file to get the extension. Is there any other way?

Comment (Saurav-D) [reply tests/test_compression.py]: Also maybe there is no need to make that method private.

File: datasetinsights/io/compression.py [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+import os
+import tarfile
+import zipfile
+
+import filetype
+
+logger = logging.getLogger(__name__)
+
+
+class ZipFileCompression:
```
Comment (AdamPalmarUnity) [inline datasetinsights/io/compression.py]: After looking around abit i found this. 

import shutil
This function could replace the different compressors.
shutil.unpack_archive() 

It takes a filename , output dir and format. If you use the _get_file_extension_from_filepath(filepath) function plus this we can saw our self from reimplementing the different formats

File: tests/test_compression.py (hunk 28 above)
Comment (AdamPalmarUnity) [reply tests/test_compression.py]: I'll push some tests and we can take a look if they make sense

File: datasetinsights/io/compression.py (hunk 29 above)
Comment (Saurav-D) [reply datasetinsights/io/compression.py]: Okay, I'll implement this.

Comment (AdamPalmarUnity) [reply datasetinsights/io/compression.py]: Remember to pull i just pushed some tests maybe they are not needed at this point, if this is implemented. Feel free to remove them if need be.

Comment (Saurav-D) [reply datasetinsights/io/compression.py]: After reading the documentation, which states: `format is the archive format: one of “zip”, “tar”, “gztar”, “bztar”, or “xztar”`. That is they only support these 5 formats rest of the formats we'll have to register. Also, out of these 5 only 2 are most popular which means apart from these 2 we'll have to anyway implement other formats. Libraries like pytorch as well have their own [implementation](https://github.com/pytorch/vision/blob/dcfcc8675393749e852d8956b580b1660dfbea2d/torchvision/datasets/utils.py#L214).

Comment (Saurav-D) [reply datasetinsights/io/compression.py]: Will implement this in new PR.

File: datasetinsights/io/download.py [hunk 30]
```diff
@@ -122,3 +124,40 @@ def _crc32_checksum(filepath):
         checksum = zlib.crc32(f.read())
 
     return checksum
+
+
+def get_checksum_from_file(filepath):
+    """ This method return checksum of the file whose filepath is given.
+
+    Args:
+        filepath (str): Path of the checksum file.
```
Comment (adason) [inline datasetinsights/io/download.py:133]: Should add note to explain filepath can be either `http url` or `local_path`

File: datasetinsights/io/download.py [hunk 31]
```diff
@@ -122,3 +124,40 @@ def _crc32_checksum(filepath):
...
+def read_checksum_from_txt(filepath):
+    """ This method reads checksum from a txt file and returns it.
+
+    Args:
+        filepath (str): Local filepath of the checksum file.
+
+    """
+    with open(filepath) as file:
+        checksum = file.read()
+    return int(checksum)
```
Comment (adason) [inline datasetinsights/io/download.py]: [mustfix] checksum isn't always integer.

File: datasetinsights/io/download.py [hunk 32]
```diff
@@ -122,3 +124,40 @@ def _crc32_checksum(filepath):
...
+    else:
+        raise ValueError(f"Can not get checksum from path: {filepath}")
+
+
+def read_checksum_from_txt(filepath):
+    """ This method reads checksum from a txt file and returns it.
+
+    Args:
+        filepath (str): Local filepath of the checksum file.
+
```
Comment (adason) [inline datasetinsights/io/download.py:162]: Missing docstring for the return value of this method? Also, is this method going to be used as public method? Maybe we can make it private?

File: datasetinsights/io/downloader/base.py [hunk 33]
```diff
@@ -12,22 +12,22 @@ def _find_downloader(source_uri):
         Args:
             source_uri: URI of where this data should be downloaded.
 
-            Returns: The dataset downloader class that is
-             registered with the source-uri protocol
-
+        Returns: The dataset downloader class that is registered with the
```
Comment (adason) [inline datasetinsights/io/downloader/base.py:15]: Return docstring should follow this pattern

```
Returns:
    type: description
```

see this [doc](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) for more examples

File: datasetinsights/io/downloader/http_downloader.py [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+    get_checksum_from_file,
+    validate_checksum,
+)
+from datasetinsights.io.downloader.base import DatasetDownloader
+from datasetinsights.io.exceptions import ChecksumError
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPDownloader(DatasetDownloader, protocol="http://"):
```
Comment (adason) [inline datasetinsights/io/downloader/http_downloader.py]: [mustfix] We agree to change this to `HTTPDatasetDownloader`

File: pyproject.toml [hunk 35]
```diff
@@ -46,6 +46,7 @@ dask = {extras = ["complete"], version = "^2.14.0"}
 dash = "1.12.0"
 tensorflow = "^2.2.0"
 click = "^7.1.2"
+filetype = "^1.0.7"
```
Comment (adason) [inline pyproject.toml]: Why is this package required in this review? Is it used as part of this review?

File: tests/test_create_downloader.py (hunk 18 above)
Comment (adason) [reply tests/test_create_downloader.py:22]: The above two cases are good to add. You will want to test both interface and implementation in unittests. We will not be able to cover all edge cases in one pass, but as we work on the repo and fixing bugs, we should continue adding new cases that will cause a bug.

**Review by adason — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 35 above)
Comment (Saurav-D) [reply pyproject.toml]: Removed.

File: datasetinsights/io/downloader/http_downloader.py (hunk 34 above)
Comment (Saurav-D) [reply datasetinsights/io/downloader/http_downloader.py]: will do.

File: datasetinsights/io/download.py (hunk 32 above)
Comment (Saurav-D) [reply datasetinsights/io/download.py:162]: We can make it private, I have one question though, should we test private methods?

**Review by adason — APPROVED** [review]
